### DISCUSSION TITLE
Encryption

### DISCUSSION THREAD

**Discussion by ceraolo** [OpenModelica/OpenModelica#11389] [Q&A]
In the Openmodelilca doumentation I read:

 _Note that you need a special version of OpenModelica with encryption support. Contact us if you want one._

I do need that special version. Who should I contact exactly?
Thank you.

**Comment by adeas31** [comment]
@adrpo is the platform for distributing encrypted version is ready or not?

**Reply by casella** [reply]
It's done but it still has issues (-> crashes). It will be ironed out together with Bosch-Rexroth once they are done with the beta testing of 1.22.0.

**Reply by ceraolo** [reply]
Great!
As soon as you have a version working at a minimal level, please consider sharing it with me.
It would be great for me to be able to do a pair of encriptions in the next days/weeks, even using a partially faulty OM.
Thank you.

**Reply by casella** [reply]
OK. Please note that encryption is available to Level 2 OSMC members.

**Reply by casella** [reply]
Of course we can set up a free demo if needed

**Reply by ceraolo** [reply]
>OK. Please note that encryption is available to Level 2 OSMC members.

Seems reasonable, but strange that is not written in the OM docu:
https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/encryption.html

> Of course we can set up a free demo if needed

Useful: before deciding to become a level 2 member to have encryption, it is wise to have the chance to test it

**Reply by ceraolo** [reply]
My need to encrypt comes from the fact that I am considering distributing modelica code along with scientific papers to let readers to reproduce the results, while hiding some core parts that I do not want to disclose.
Not for commercial purposes.

Given this, I may consider becoming Level 2, only in case the reader can access encrypted code using standard OpenModelica (I mean without having to become OSMC member at all). Is this the case?

**Reply by casella** [reply]
If you become a Level 2 member, you can redistribute the version of OMC with encryption support to all of your users/customers, or have them download it from OSMC website on your behalf. Up to 10 if you are a limited level 2 member, no limitations for unlimited membership. According to the current bylaws, if your group is less than 10 people, limited level 2 is 2200 €/year, unlimited level 2 is 3300 €/year.

**Reply by ceraolo** [reply]
Quite clear.

So, ordinary OM cannot nor create nor read mol files. It was a bit confusing that its file menu contains, non-grayed, the item "load Encrypted library"

Therefore, would I publish on github twin modelica code of scientific papers, containing some encrypted parts, I should have an unlimited level2 membership.

I still see something strange on this mechanism. Would I have unlimited level2, I could let all of my users to download OM with encryption from OSMC site, therefore virtually killing any possibility of further income for OSMC from that executable because access with OM with encryption would have become public!
One entity (the level2 unlimited customer) pays for all mankind to possibly use OM with encryption... There must be something I don't understand!


Thanks.

**Comment by casella** [comment]
> So, ordinary OM cannot nor create nor read mol files. It was a bit confusing that its file menu contains, non-grayed, the item "load Encrypted library"

That is a bug, not a feature 😄 See #11437.

> Therefore, would I publish on github twin modelica code of scientific papers, containing some encrypted parts, I should have an unlimited level2 membership.

This is how it works now.

> I still see something strange on this mechanism. Would I have unlimited level2, I could let all of my users to download OM with encryption from OSMC site, therefore virtually killing any possibility of further income for OSMC from that executable because access with OM with encryption would have become public! One entity (the level2 unlimited customer) pays for all mankind to possibly use OM with encryption... There must be something I don't understand!

This process requires registration, so if the whole mankind starts downloading encrypted OMC on your behalf we would notice. 

When we start seeing a serious number of downloads, we can also add some extra mechanism to the website by which one would need to input some unique token that _you_ generated in order to download the version on your behalf. Maybe we already have that in place, I'm not sure about the latest developments. That would be an additional burden for you, so I would do it only if necessary.

Of course it is always possible that someone downloads the version of OpenModelica with encryption support from the website, and then uploads it on some offshore wharez website. But that could actually be a version of OpenModelica that they compiled, including some bad malware that would screw your whole system up instantly. If you want to run that risk, that is up to you 😄. I would only download OpenModelica from a trusted source (e.g. the openmodelica.org website).

At the end of the day, although the OSMC's business model includes the possibility of some non-open-source activity, we fundamentally remain an open-source organization. Dealing with elaborate copy-protection schemes or licensing software such as flexlm (for which we would have to pay for) is really beyond our scope.

**Reply by ceraolo** [reply]
As an OSMC member I would suggest considering a paradigm shift:
- restricting OM with mol _creation_  to  selected users (e.g. as it is now: level 2 members)
- allowing _reading_ (and executing) mol files in ordinary OM

Frankly speaking, this scheme seems to me much cleaner.

_At the end of the day, although the OSMC's business model includes the possibility of some non-open-source activity, we fundamentally remain an open-source organization._ 

The user model for mol files I have in mind goes exactly in this direction. Consider what I said:

_My need to encrypt comes from the fact that I am considering distributing modelica code along with scientific papers to let readers to reproduce the results, while hiding some core parts that I do not want to disclose.
Not for commercial purposes._

So, my usage model was to **enlarge** the usage and power of open-sourceness of scientific world, not restrict it.

****************
And... if this is accepted, my university would gladly accept complimentary access to mol writing, just as an acknoledgement for its cotribution to this paradigm shift  ;-)

**Reply by adeas31** [reply]
Adding @arun3688 in this discussion. He is working on the distribution functionality.

**Comment by arun3688** [comment]
@adeas31 the website for distributing encrypted version is ready and me and adrian made some test, but right how it is offline, adrian will have to restart the machine i guess

**Comment by casella** [comment]
> As an OSMC member I would suggest considering a paradigm shift:
> 
>     * restricting OM with mol _creation_  to  selected users (e.g. as it is now: level 2 members)
> 
>     * allowing _reading_ (and executing) mol files in ordinary OM

This is much easier on the user's side, but poses two potential problems:
- we have no way to enforce the 10-users max limit for Level 2-limited OSMC members. In fact, if the OSMC member distributes the encryped OpenModelica version himself, we don't have that either, so it's already based on good faith
- I have to check the specification SEMLA protocol in detail, see it [here](https://github.com/modelon-community/SEMLA/blob/master/doc/SEMLA.md), to see if it could support it. Normally, encryption is made with a public key and decryption with the private key, and the private key cannot be put in the regular open-source OMC. Maybe we could swap that, but I'm not sure if the protocol allows it

We could have a discussion about it after the devmeeting next Monday.

> And... if this is accepted, my university would gladly accept complimentary access to mol writing, just as an acknowledgement for its contribution to this paradigm shift ;-)

In-kind contributions are always welcome :)

**Comment by ceraolo** [comment]
>we have no way to enforce the 10-users max limit for Level 2-limited OSMC members. In fact, if the OSMC member distributes the encryped OpenModelica version himself, we don't have that either, so it's already based on good faith

For mol writing the situation wil not change in comparison with now: based on food faith.
What changes is that everyone would become able to mol reading. This is huge change, but can possibly bring more advantages than disadvantages. Consider that after one company has got level2 unlimited, it would be probable that OM with encryption distribution gets out of control. With my proposal, reading mols will be possible for everyone, but writing much more under control. 

> * I have to check the specification SEMLA protocol in detail, see it [here](https://github.com/modelon-community/SEMLA/blob/master/doc/SEMLA.md), to see if it could support it. Normally, encryption is made with a public key and decryption with the private key, and the private key cannot be put in the regular open-source OMC. Maybe we could swap that, but I'm not sure if the protocol allows it

Obviously, the published OM code should be deprived of the decryption part. Those compiling it themselves will create executables without decryption
> 
> We could have a discussion about it after the devmeeting next Monday.

Thanks a lot.

**Comment by casella** [comment]
@ceraolo we further discussed the issue during today's webmeeting. Apparently, there is a catch that we missed: anyone can encrypt a library using the public key and the SEMLA protocol, _you don't need to use OMC to do that_. If we then release the version with decryption support to the general public, we lose one of the motivations for organizations to join as Level 2 members.

This may still be OK, but we have to think about it.

**Comment by ceraolo** [comment]
Aha, I also did not realise this at all.
I'll wait for your final decisions

**Comment by ceraolo** [comment]
Francesco, ho a uto la conferma che il pf di Terna sceglie sempre per il saldo un nodo in cui la potenza (attiva e reattiva) possibilmente generabile sia ingente.Quindi quest'ultimo sforzo che vi sto facendo fare ha un senso.
